File: src/test/java/com/mbayou/kick4k/events/KickEventDispatcherTest.java
package com.mbayou.kick4k.events;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import org.junit.jupiter.api.BeforeEach;
import org.mockito.ArgumentCaptor;
import com.mbayou.kick4k.events.handler.KickEventDispatcher;
import com.mbayou.kick4k.events.handler.KickEventListener;
import com.mbayou.kick4k.events.type.KickEvent;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

public class KickEventDispatcherTest {
    protected KickEventDispatcher eventDispatcher;

    @BeforeEach
    public void setup() {
        ObjectMapper objectMapper = new ObjectMapper()
                .setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE)
                .registerModule(new JavaTimeModule());

        this.eventDispatcher = new KickEventDispatcher(objectMapper);
    }

    protected <E extends KickEvent> E dispatchAndCapture(
            Class<E> eventClass, String eventType, String version, String payload) {

        KickEventListener<E> listener = mock(KickEventListener.class);
        this.eventDispatcher.registerListener(eventClass, listener);

        this.eventDispatcher.dispatch(eventType, version, payload);

        ArgumentCaptor<E> captor = ArgumentCaptor.forClass(eventClass);
        verify(listener, times(1)).handle(captor.capture());

        return captor.getValue();
    }
}

File: src/test/kotlin/com/mbayou/kick4k/events/KickEventDispatcherTest.kt
package com.mbayou.kick4k.events

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.PropertyNamingStrategies
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.mbayou.kick4k.events.handler.KickEventDispatcher
import com.mbayou.kick4k.events.handler.KickEventListener
import com.mbayou.kick4k.events.type.KickEvent
import org.junit.jupiter.api.BeforeEach

open class KickEventDispatcherTest {
    protected lateinit var eventDispatcher: KickEventDispatcher

    @BeforeEach
    fun setup() {
        val objectMapper = ObjectMapper()
            .setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE)
            .registerModule(JavaTimeModule())
        eventDispatcher = KickEventDispatcher(objectMapper)
    }

    protected fun <E : KickEvent> dispatchAndCapture(
        eventClass: Class<E>,
        eventType: String,
        version: String,
        payload: String,
    ): E {
        val events = mutableListOf<E>()
        val listener = KickEventListener<E> { event -> events.add(event) }
        eventDispatcher.registerListener(eventClass, listener)
        eventDispatcher.dispatch(eventType, version, payload)
        return events.first()
    }
}
